Ethan Ho  
CMPE 256  
May 3, 2026  
HW8 Graphs  
NEtworkX Library

In [2]:
# !curl -o wiki-Vote.txt.gz https://snap.stanford.edu/data/wiki-Vote.txt.gz
# !gunzip wiki-Vote.txt.gz

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  283k  100  283k    0     0   282k      0  0:00:01  0:00:01 --:--:--  282k


In [36]:
import networkx as nx
from community import community_louvain as co

In [7]:
G = nx.DiGraph()

with open('wiki-Vote.txt', 'r') as f:
    for line in f:
        if line.startswith('#'):
            continue
        u, v = map(int, line.strip().split())
        G.add_edge(u, v)

In [8]:
print(f"Number of nodes: {G.number_of_nodes()}")
print(f"Number of edges: {G.number_of_edges()}")

Number of nodes: 7115
Number of edges: 103689


1: Number of nodes and number of edges [1pt]

There are 7115 nodes and 103689 edges in the graph.

In [13]:
is_strongly_connected = nx.is_strongly_connected(G)
print(f"The graph is {is_strongly_connected} connected")

if is_strongly_connected:
    num_components = nx.number_of_strongly_connected_components(G)
    print(f"{num_components} connected components")
else:
    num_components = nx.number_weakly_connected_components(G)
    print(f"{num_components} connected components")

The graph is False connected
24 connected components


2: Is the graph strongly connected? If yes, return the number of strongly connected components. If not, return the number of weakly connected components.  [1pt]

The graph is a weakly connected graph with 24 weak connections

In [19]:
longest_shortest_path = 0
for component_nodes in nx.weakly_connected_components(G):
    component_subgraph = G.subgraph(component_nodes)
    undirected_component = component_subgraph.to_undirected()

    if nx.is_connected(undirected_component) and len(undirected_component) > 1:
          component_diameter = nx.diameter(undirected_component)
          longest_shortest_path = max(longest_shortest_path, component_diameter)
    elif len(undirected_component) == 1:
        pass

longest_shortest_path

7

3: The diameter of the graph (longest shortest path) -- hint:think of what needs to be reported if your graph is not connected? [1pt]

The diameter of the graph is 7.

In [20]:
nx.density(G)

0.0020485375110809584

4: The density of the graph. [0.5pt]

The density of the graph is 0.0020485375110809584

In [21]:
nx.average_clustering(G)

0.08156344522820935

5: The average clustering coefficient of the graph. [0.5pt]

The average clustering coefficient of the graph is 0.08156344522820935

In [22]:
closeness_centrality = nx.closeness_centrality(G)

sorted_closeness = sorted(closeness_centrality.items(), key=lambda item: item[1], reverse=True)
sorted_closeness[:2]

[(4037, 0.29648297322467565), (15, 0.29148957578089163)]

6a:  The two nodes with the highest closeness centrality [0.5pt]

The two nodes with the highest closeness centrality would be 4037 and 15 with their respective centralities of 0.29648297322467565 and 0.29148957578089163


In [23]:
betweenness_centrality = nx.betweenness_centrality(G)

sorted_betweenness = sorted(betweenness_centrality.items(), key=lambda item: item[1], reverse=True)
sorted_betweenness[:2]

[(2565, 0.017654409558147836), (1549, 0.016564095998753692)]

6b: The two nodes with the highest betweenness centrality [0.5pt]

The two nodes with the highest beweenness centrality is 2565 and 1549 with their respective centrailities of 0.017654409558147836 and 0.016564095998753692

In [24]:
eigenvector_centrality = nx.eigenvector_centrality(G, max_iter=1000)

sorted_eigenvector = sorted(eigenvector_centrality.items(), key=lambda item: item[1], reverse=True)
sorted_eigenvector[:2]

[(2398, 0.1171983832471754), (4037, 0.10895543930803385)]

6c: The two nodes with the highest eigenvector centrality [0.5pt] (Eigenvector centrality is another centrality metric often used)

The two nodes with the highest beweenness centrality is 2398 and 4037 with their respective centrailities of 0.1171983832471754 and 0.10895543930803385

In [31]:
pagerank = nx.pagerank(G)

sorted_pagerank = sorted(pagerank.items(), key=lambda item: item[1], reverse=True)
sorted_pagerank[:2]

[(4037, 0.004612715891167545), (15, 0.0036812207295292714)]

6d: The two nodes with the highest PageRank [0.5pt]

The two nodes with the highest pagerank would be 4037 and 15

In [26]:
hubs, authorities = nx.hits(G)

sorted_hubs = sorted(hubs.items(), key=lambda item: item[1], reverse=True)
sorted_authorities = sorted(authorities.items(), key=lambda item: item[1], reverse=True)

print(sorted_hubs[:2])
print(sorted_authorities[:2])

[(2565, 0.007940492708143138), (766, 0.007574335297501241)]
[(2398, 0.0025801471780088747), (4037, 0.002573241124229799)]


6e: The two nodes with the highest HITS hub [0.5pt] and HITS authority [0.5pt] score.

The two nodes with the highest HITS hub would be 2565 and 766 while the two nodes with the highest authority would be 2398 and 4037

6f: What do you observe? Are there overlaps among the retrieved pairs? [0.5 pt]

I can see that a few nodes stand out to be qite important when rated among even three centrality methods. Those nodes with overlap would be 2565, 2398, 4037, and 15

In [30]:
node1_pr_data, node2_pr_data = sorted_pagerank[0], sorted_pagerank[1]
node1_pr, node2_pr = node1_pr_data[0], node2_pr_data[0]

jaccard_similarity_pr = list(nx.jaccard_coefficient(G.to_undirected(), ebunch=[(node1_pr, node2_pr)]))[0][2]

print(f"Jacc Sim between {node1_pr} and {node2_pr} is {jaccard_similarity_pr}")

Jacc Sim between 4037 and 15 is 0.14323258869908015


7a: The Jaccard similarity between the two nodes with highest PageRank [0.5pt]

The Jaccard similarity between the two nodes with the highest page rank is 0.14323258869908015

In [33]:
node1_cc_data, node2_cc_data = sorted_closeness[0], sorted_closeness[1]
node1_cc, node2_cc = node1_cc_data[0], node2_cc_data[0]

jaccard_similarity_cc = list(nx.jaccard_coefficient(G.to_undirected(), ebunch=[(node1_cc, node2_cc)]))[0][2]

print(f"Jacc Sim between {node1_cc} and {node2_cc} is {jaccard_similarity_cc}")

Jacc Sim between 4037 and 15 is 0.14323258869908015


7b: The Jaccard similarity between the two nodes with highest closeness centrality [0.5pt]

The Jaccard similarity between the two nodes with the highest closeness centrality is 0.14323258869908015.


7c: What can you observe ? [0.5 pt]

I can see that the two jaccard similarities are exaclty the same dispite using different algorithms.

In [37]:
undirected_G = G.to_undirected()

partition = co.best_partition(undirected_G)

len(set(partition.values()))

32

8: Find the communities using a community detection algorithm of your choice. How many communities does your algorithm find? (even for hierarchical algorithms there's usually an optimal split). [1pt]     

I decided to use the Louvian algorthm since it ran in Linear time compared to Girvan-Newman. I used a libary called community with Louvian for simplicity and got 32 communities from our graph